### PR TITLE
new 03 Groups 

### PR DESCRIPTION
**Pull request by Bilb**

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -5,6 +5,7 @@ on:
     branches:
       - main
       - dev
+      - groups-chunk-2
```
Comment (yougotwill) [inline .github/workflows/test.yml]: What are we testing here?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -13,8 +14,11 @@ concurrency:
 jobs:
   build:
     runs-on: ${{ matrix.os }}
+
     strategy:
       fail-fast: false
+
```
Comment (yougotwill) [inline .github/workflows/test.yml]: Cleanup new lines?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -47,6 +61,14 @@ jobs:
...
+        shell: bash
+        run: |
+          sed -i "s/target_compile_options(oxen-logging-warnings INTERFACE/#target_compile_options(oxen-logging-warnings INTERFACE/" libsession-util/external/oxen-libquic/external/oxen-logging/CMakeLists.txt
+          cat libsession-util/external/oxen-libquic/external/oxen-logging/CMakeLists.txt
+
+
       - name: build libsession-util-nodejs
         shell: bash
-        run: yarn install --frozen-lockfile --network-timeout 600000
+        run:  yarn install --frozen-lockfile
```
Comment (yougotwill) [inline .github/workflows/test.yml]: extra space?

File: src/base_config.hpp [hunk 4]
```diff
@@ -104,17 +106,18 @@ class ConfigBaseImpl {
 
             Napi::Env env = info.Env();
 
-            config->logger = [env, class_name](session::config::LogLevel, std::string_view x) {
-                std::string toLog =
-                        "libsession-util:" + std::string(class_name) + ": " + std::string(x) + "\n";
+            // config->logger = [env, class_name](session::config::LogLevel, std::string_view x) {
```
Comment (yougotwill) [inline src/base_config.hpp]: To be used later?

File: src/groups/meta_group.hpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+using session::config::profile_pic;
+using std::pair;
+using std::string;
+using std::tuple;
+using std::vector;
+
+using std::make_shared;
+using std::shared_ptr;
+
+class MetaGroup {
```
Comment (yougotwill) [inline src/groups/meta_group.hpp:23]: Why call it MetaGroup? Prevent any conflicts with `libsession`?

File: src/groups/meta_group_wrapper.cpp [hunk 6]
```diff
@@ -0,0 +1,781 @@
...
+        Napi::ObjectWrap<MetaGroupWrapper>{info} {}
+
+void MetaGroupWrapper::Init(Napi::Env env, Napi::Object exports) {
+    MetaBaseWrapper::NoBaseClassInitHelper<MetaGroupWrapper>(
+            env,
+            exports,
+            "MetaGroupWrapperNode",
+            {
+                    // shared exposed functions
+
```
Comment (yougotwill) [inline src/groups/meta_group_wrapper.cpp]: Delete new line please

File: src/groups/meta_group_wrapper.cpp [hunk 7]
```diff
@@ -0,0 +1,781 @@
...
+                    InstanceMethod("infoGet", &MetaGroupWrapper::infoGet),
+                    InstanceMethod("infoSet", &MetaGroupWrapper::infoSet),
+                    InstanceMethod("infoDestroy", &MetaGroupWrapper::infoDestroy),
+
+                    // members exposed functions
+                    InstanceMethod("memberGet", &MetaGroupWrapper::memberGet),
+                    InstanceMethod("memberGetOrConstruct", &MetaGroupWrapper::memberGetOrConstruct),
+                    InstanceMethod(
+                            "memberConstructAndSet", &MetaGroupWrapper::memberConstructAndSet),
+
```
Comment (yougotwill) [inline src/groups/meta_group_wrapper.cpp]: Delete new line

File: src/groups/meta_group_wrapper.cpp [hunk 8]
```diff
@@ -0,0 +1,781 @@
...
+                            &MetaGroupWrapper::memberSetPromotionFailed),
+                    InstanceMethod(
+                            "memberSetPromotionAccepted",
+                            &MetaGroupWrapper::memberSetPromotionAccepted),
+                    InstanceMethod(
+                            "memberSetProfilePicture", &MetaGroupWrapper::memberSetProfilePicture),
+                    InstanceMethod("memberEraseAndRekey", &MetaGroupWrapper::memberEraseAndRekey),
+
+                    // keys exposed functions
+
```
Comment (yougotwill) [inline src/groups/meta_group_wrapper.cpp]: Delete new line

File: src/groups/meta_group_wrapper.cpp [hunk 9]
```diff
@@ -0,0 +1,781 @@
...
+/* #region INFO ACTIONS */
+
+Napi::Value MetaGroupWrapper::infoGet(const Napi::CallbackInfo& info) {
+    return wrapResult(info, [&] {
+        auto env = info.Env();
+        auto obj = Napi::Object::New(env);
+
+        obj["name"] = toJs(env, this->meta_group->info->get_name());
+        obj["createdAtSeconds"] = toJs(env, this->meta_group->info->get_created());
+        obj["deleteAttachBeforeSeconds"] =
```
Comment (yougotwill) [inline src/groups/meta_group_wrapper.cpp:322]: What is `deleteAttachBeforeSeconds`?

File: src/groups/meta_group_wrapper.cpp [hunk 10]
```diff
@@ -0,0 +1,781 @@
...
+        return allMembersRemoved;
+    });
+}
+
+Napi::Value MetaGroupWrapper::memberGet(const Napi::CallbackInfo& info) {
+    return wrapResult(info, [&] {
+        assertInfoLength(info, 1);
+        assertIsString(info[0]);
+
+        auto pubkeyHex = toCppString(info[0], __PRETTY_FUNCTION__);
```
Comment (yougotwill) [inline src/groups/meta_group_wrapper.cpp]: Want 🧙‍♂️is `__PRETTY_FUNCTION__`?

File: src/groups/meta_group_wrapper.cpp [hunk 11]
```diff
@@ -0,0 +1,781 @@
...
+        auto m = this->meta_group->members->get(pubkeyHex);
+        if (m) {
+            m->set_promotion_accepted();
+            this->meta_group->members->set(*m);
+        }
+    });
+}
+
+void MetaGroupWrapper::memberSetProfilePicture(const Napi::CallbackInfo& info) {
+
```
Comment (yougotwill) [inline src/groups/meta_group_wrapper.cpp]: Delete new line

File: src/groups/meta_group_wrapper.cpp [hunk 12]
```diff
@@ -0,0 +1,781 @@
...
+        assertIsUInt8Array(info[0], "swarmSubaccountSign 0");
+        assertIsUInt8Array(info[1], "swarmSubaccountSign 1");
+
+        auto message = toCppBuffer(info[0], __PRETTY_FUNCTION__);
+        auto authdata = toCppBuffer(info[1], __PRETTY_FUNCTION__);
+        auto subaccountSign = this->meta_group->keys->swarm_subaccount_sign(message, authdata);
+
+        return subaccountSign;
+    });
+}
```
Comment (yougotwill) [inline src/groups/meta_group_wrapper.cpp:772]: Please split up the actions into separate files and import if possible.

File: src/groups/meta_group_wrapper.hpp [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        } else if (info.admin) {
+            obj["memberStatus"] = toJs(env, "PROMOTION_NOT_SENT");
+        } else if (info.invite_status == 0) {
+            obj["memberStatus"] = toJs(env, "INVITE_ACCEPTED");
+        } else if (info.invite_not_sent()) {
+            obj["memberStatus"] = toJs(env, "INVITE_NOT_SENT");
+        } else if (info.invite_failed()) {
+            obj["memberStatus"] = toJs(env, "INVITE_FAILED");
+        } else if (info.invite_pending()) {
+            // Note: INVITE_NOT_SENT is 3, which makes invite_pending() return true, so be sure to
```
Comment (yougotwill) [inline src/groups/meta_group_wrapper.hpp:47]: Do you mean '3' as in a protobuf or enum? I don't understand this 😅

File: src/groups/meta_group_wrapper.hpp [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+    void memberSetPromoted(const Napi::CallbackInfo& info);
+    void memberSetPromotionSent(const Napi::CallbackInfo& info);
+    void memberSetPromotionFailed(const Napi::CallbackInfo& info);
+    void memberSetPromotionAccepted(const Napi::CallbackInfo& info);
+    void memberSetProfilePicture(const Napi::CallbackInfo& info);
+    Napi::Value memberEraseAndRekey(const Napi::CallbackInfo& info);
+    void membersMarkPendingRemoval(const Napi::CallbackInfo& info);
+
+    /** Keys Actions */
+
```
Comment (yougotwill) [inline src/groups/meta_group_wrapper.hpp]: Delete new line

File: src/meta/meta_base_wrapper.hpp [hunk 15]
```diff
@@ -29,6 +30,82 @@ class MetaBaseWrapper {
...
+
+            std::optional<ustring> dumped_meta = maybeNonemptyBuffer(
+                    obj.Get("metaDumped"), class_name + ":constructGroupWrapper.metaDumped");
+
+            std::optional<ustring_view> dumped_info;
+            std::optional<ustring_view> dumped_members;
+            std::optional<ustring_view> dumped_keys;
+
+            if (dumped_meta) {
+
```
Comment (yougotwill) [inline src/meta/meta_base_wrapper.hpp]: Delete new line

File: src/multi_encrypt/multi_encrypt.hpp [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+#include <algorithm>
+
+#include "session/config/user_profile.hpp"
+#include "session/multi_encrypt.hpp"
+#include "session/random.hpp"
+#include "../utilities.hpp"
+
+namespace session::nodeapi {
+
```
Comment (yougotwill) [inline src/multi_encrypt/multi_encrypt.hpp:13]: Delete new line

File: src/multi_encrypt/multi_encrypt.hpp [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+class MultiEncryptWrapper : public Napi::ObjectWrap<MultiEncryptWrapper> {
+
```
Comment (yougotwill) [inline src/multi_encrypt/multi_encrypt.hpp]: Delete new line

File: src/multi_encrypt/multi_encrypt.hpp [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+  public:
+    MultiEncryptWrapper(const Napi::CallbackInfo& info) :
+            Napi::ObjectWrap<MultiEncryptWrapper>{info} {
+        throw std::invalid_argument(
+                "MultiEncryptWrapper is all static and don't need to be constructed");
```
Comment (yougotwill) [inline src/multi_encrypt/multi_encrypt.hpp]: "MultiEncryptWrapper is all static and don't need to be constructed" -> "MultiEncryptWrapper is static and doesn't need to be constructed"

File: types/user/usergroups.d.ts [hunk 19]
```diff
@@ -1,40 +1,62 @@
...
+  };
+
+  export type LegacyGroupInfo = BaseUserGroup & {
     pubkeyHex: string; // The legacy group "session id" (33 bytes).
     name: string; // human-readable; this should normally always be set, but in theory could be set to an empty string.
-    encPubkey: Uint8Array; // bytes (32 or empty)
+    encPubkey: Uint8ArrayLen32; // bytes (32 or empty)
     encSeckey: Uint8Array; // bytes (32 or empty)
-    priority: number; // -1 means hidden, 0 means normal, > 1 means pinned. We currently don't support hidden groups on the client though
+    // disappearingTimerSeconds: number; // in seconds, 0 == disabled.
```
Comment (Bilb) [inline types/user/usergroups.d.ts]: this might be an issue on my side

File: src/multi_encrypt/multi_encrypt.hpp [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+                assertIsUInt8Array(itemValue, "multiEncrypt.itemValue.recipient");
+                auto item = toCppBuffer(itemValue, "multiEncrypt.itemValue.recipient");
+                recipients.push_back(item);
+            }
+            ustring random_nonce = session::random::random(24);
+
+            std::vector<ustring_view> messages_sv(messages.begin(), messages.end());
+            std::vector<ustring_view> recipients_sv(recipients.begin(), recipients.end());
+
+            // Note: this function needs the first 2 args to be vector of sv explicitely
```
Comment (yougotwill) [inline src/multi_encrypt/multi_encrypt.hpp]: sv explicitely -> sv explicitly

File: types/user/usergroups.d.ts (hunk 19 above)
Comment (Bilb) [reply types/user/usergroups.d.ts]: actually no, I should just remove that line, as it is now part of `type BaseUserGroup`

File: src/user_config.cpp [hunk 21]
```diff
@@ -1,11 +1,15 @@
...
+#include <iostream>
+
 #include "base_config.hpp"
 #include "profile_pic.hpp"
+#include "session/config/base.hpp"
 #include "session/config/user_profile.hpp"
 
 namespace session::nodeapi {
 
+// using config::LogLevel;
```
Comment (yougotwill) [inline src/user_config.cpp]: Cleanup?

File: src/utilities.cpp [hunk 22]
```diff
@@ -168,4 +190,40 @@ int64_t unix_timestamp_now() {
...
+Napi::Object push_result_to_JS(
+        const Napi::Env& env,
+        const push_entry_t& push_entry,
+        const session::config::Namespace& push_namespace) {
+    auto obj = Napi::Object::New(env);
+
+    obj["seqno"] = toJs(env, std::get<0>(push_entry));
+    obj["data"] = toJs(env, std::get<1>(push_entry));
+    obj["hashes"] = toJs(env, std::get<2>(push_entry));
+    obj["namespace"] = toJs(env, push_namespace);
```
Comment (yougotwill) [inline src/utilities.cpp:202]: Will "results" always have these props?

File: src/utilities.hpp [hunk 23]
```diff
@@ -122,6 +140,20 @@ struct toJs_impl<std::vector<T>> {
...
+
+template <typename T>
+struct toJs_impl<std::unordered_set<T>> {
+    auto operator()(const Napi::Env& env, const std::unordered_set<T>& set) {
+        std::vector<T> as_array(set.begin(), set.end());
+
+        auto arr = Napi::Array::New(env, as_array.size());
+        for (size_t i = 0; i < as_array.size(); i++)
+            arr[i] = toJs(env, as_array[i]);
+
```
Comment (yougotwill) [inline src/utilities.hpp]: Delete new line

File: types/groups/groupmembers.d.ts [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+    memberGetOrConstruct: (pubkeyHex: PubkeyType) => GroupMemberGet;
+    memberConstructAndSet: (pubkeyHex: PubkeyType) => void;
+
+    memberGetAll: () => Array<GroupMemberGet>;
+    memberGetAllPendingRemovals: () => Array<GroupMemberGet>;
+
+    // setters
+    memberSetNameTruncated: (pubkeyHex: PubkeyType, newName: string) => void;
+
+    /** A member invite states defaults to invite-not-sent. Use this function to mark that you've sent one, or at least tried (failed: boolean)*/
```
Comment (yougotwill) [inline types/groups/groupmembers.d.ts]: Maybe "A member invite states defaults to" -> "A member's invite state defaults to"

File: types/groups/index.d.ts [hunk 25]
```diff
@@ -0,0 +1,2 @@
+/// <reference path="./groupinfo.d.ts" />
```
Comment (yougotwill) [inline types/groups/index.d.ts]: Might be a noob question but if this is already referenced in `metagroup.d.ts` does it still need to be here?

File: types/groups/metagroup.d.ts [hunk 26]
```diff
@@ -0,0 +1,156 @@
+/// <reference path="../shared.d.ts" />
+/// <reference path="./groupmembers.d.ts" />
+/// <reference path="./groupinfo.d.ts" />
+/// <reference path="./groupkeys.d.ts" />
+
+declare module 'libsession_util_nodejs' {
+  export type ConfirmKeysPush = [data: Uint8Array, hash: string, timestampMs: number];
+
+  export type GroupWrapperConstructor = {
+    userEd25519Secretkey: Uint8Array; // (len 64)
```
Comment (yougotwill) [inline types/groups/metagroup.d.ts]: docstring comments above would be nice so we get this info on hover in our ide?

File: types/shared.d.ts [hunk 27]
```diff
@@ -0,0 +1,162 @@
+declare module 'libsession_util_nodejs' {
+  type Uint8ArrayFixedLength<T extends number> = {
+    buffer: Uint8Array;
+    length: T;
+  };
+  /**
+   * Allow a single type to be Nullable. i.e. string => string | null
+   */
+  type Nullable<T> = T | null;
```
Comment (yougotwill) [inline types/shared.d.ts]: Might be good to export Nullable for use in session desktop? Same for AllFieldsNullable

File: types/user/usergroups.d.ts [hunk 28]
```diff
@@ -1,40 +1,62 @@
...
     members: Array<LegacyGroupMemberInfo>;
-    joinedAtSeconds: number; // equivalent to the lastJoinedTimestamp in Session desktop but in seconds rather than MS
-    // NOTE Properties need to be optional going forward to backward compatibility
-    disappearingTimerSeconds?: number; // in seconds, 0 or undefined == disabled.
   };
 
+  type UserGroupsGet = BaseUserGroup & {
+    pubkeyHex: GroupPubkeyType; // The group "session id" (33 bytes), starting with 03.
```
Comment (yougotwill) [inline types/user/usergroups.d.ts]: Docstrings above for ide hover

**Review by yougotwill — CHANGES_REQUESTED** [review]

File: src/groups/meta_group.hpp (hunk 5 above)
Comment (Bilb) [reply src/groups/meta_group.hpp:23]: yes

Comment (Bilb) [reply src/groups/meta_group.hpp:23]: we have usergroups on one side, and GroupMembers, GroupInfos, GroupKeys on the other one. This MetaGroup unifies GroupMembers, GroupInfos, GroupKeys

File: .github/workflows/test.yml (hunk 3 above)
Comment (Bilb) [reply .github/workflows/test.yml]: :+1:

File: src/base_config.hpp (hunk 4 above)
Comment (Bilb) [reply src/base_config.hpp]: no, latest libsesion-util broke it and it won't be as easy to use it back. Will remove this in the meantime.

File: src/groups/meta_group_wrapper.cpp (hunk 9 above)
Comment (Bilb) [reply src/groups/meta_group_wrapper.cpp:322]: delete attachments before this timestamp in seconds. A future setting we will be able to add as an admin: all members will delete the messages sent before X, which had attachments in them,

File: src/groups/meta_group_wrapper.cpp (hunk 10 above)
Comment (Bilb) [reply src/groups/meta_group_wrapper.cpp]: it's a C macro which grabs the name of the current function as a string.

File: src/groups/meta_group_wrapper.cpp (hunk 12 above)
Comment (Bilb) [reply src/groups/meta_group_wrapper.cpp:772]: not possible, and not ideal. Besides conversation.ts is 3000 lines long on session-desktop

File: src/groups/meta_group_wrapper.hpp (hunk 13 above)
Comment (Bilb) [reply src/groups/meta_group_wrapper.hpp:47]: as in the info.member_status field
```cpp
enum groups_members_invite_status { INVITE_SENT = 1, INVITE_FAILED = 2, INVITE_NOT_SENT = 3 };
enum groups_members_remove_status { REMOVED_MEMBER = 1, REMOVED_MEMBER_AND_MESSAGES = 2 };
```

File: src/multi_encrypt/multi_encrypt.hpp (hunk 16 above)
Comment (Bilb) [reply src/multi_encrypt/multi_encrypt.hpp:13]: this one is always there	
image
not removed

File: src/multi_encrypt/multi_encrypt.hpp (hunk 17 above)
Comment (Bilb) [reply src/multi_encrypt/multi_encrypt.hpp]: removed

File: src/utilities.cpp (hunk 22 above)
Comment (Bilb) [reply src/utilities.cpp:202]: you mean `push_entry` or `obj`?

File: types/groups/index.d.ts (hunk 25 above)
Comment (Bilb) [reply types/groups/index.d.ts]: maybe not, I've removed it.

File: types/shared.d.ts (hunk 27 above)
Comment (Bilb) [reply types/shared.d.ts]: exported

File: types/user/usergroups.d.ts (hunk 28 above)
Comment (Bilb) [reply types/user/usergroups.d.ts]: added some/

File: types/groups/metagroup.d.ts (hunk 26 above)
Comment (Bilb) [reply types/groups/metagroup.d.ts]: done for the constructor

File: .github/workflows/test.yml (hunk 1 above)
Comment (Bilb) [reply .github/workflows/test.yml]: yeah, I need to get some sort of CI going, will do it now.

File: src/groups/meta_group_wrapper.cpp (hunk 9 above)
Comment (yougotwill) [reply src/groups/meta_group_wrapper.cpp:322]: Cool

File: src/groups/meta_group_wrapper.cpp (hunk 10 above)
Comment (yougotwill) [reply src/groups/meta_group_wrapper.cpp]: Noice

File: src/groups/meta_group_wrapper.cpp [hunk 29]
```diff
@@ -0,0 +1,774 @@
...
+Napi::Value MetaGroupWrapper::memberGetAll(const Napi::CallbackInfo& info) {
+    return wrapResult(info, [&] {
+        std::vector<session::config::groups::member> allMembers;
+        for (auto& member : *this->meta_group->members) {
+            allMembers.push_back(member);
+        }
+        return allMembers;
+    });
+}
+Napi::Value MetaGroupWrapper::memberGetAllPendingRemovals(const Napi::CallbackInfo& info) {
```
Comment (yougotwill) [inline src/groups/meta_group_wrapper.cpp:395]: MIssing a new line here

File: src/groups/meta_group_wrapper.cpp (hunk 12 above)
Comment (yougotwill) [reply src/groups/meta_group_wrapper.cpp:772]: All good

File: src/groups/meta_group_wrapper.hpp (hunk 13 above)
Comment (yougotwill) [reply src/groups/meta_group_wrapper.hpp:47]: Cool

File: src/multi_encrypt/multi_encrypt.hpp (hunk 16 above)
Comment (yougotwill) [reply src/multi_encrypt/multi_encrypt.hpp:13]: Cool.

File: types/groups/metagroup.d.ts [hunk 30]
```diff
@@ -0,0 +1,169 @@
...
+    public memberSetPromotionAccepted: MetaGroupWrapper['memberSetPromotionAccepted'];
+    public memberSetPromotionSent: MetaGroupWrapper['memberSetPromotionSent'];
+    public memberSetPromotionFailed: MetaGroupWrapper['memberSetPromotionFailed'];
+    public memberSetInvited: MetaGroupWrapper['memberSetInvited'];
+    public memberEraseAndRekey: MetaGroupWrapper['memberEraseAndRekey'];
+    public membersMarkPendingRemoval: MetaGroupWrapper['membersMarkPendingRemoval'];
+    public memberSetProfilePicture: MetaGroupWrapper['memberSetProfilePicture'];
+
+    // keys
+
```
Comment (yougotwill) [inline types/groups/metagroup.d.ts]: Delete new line

File: types/groups/metagroup.d.ts [hunk 31]
```diff
@@ -0,0 +1,169 @@
...
+    public keyGetCurrentGen: MetaGroupWrapper['keyGetCurrentGen'];
+    public encryptMessages: MetaGroupWrapper['encryptMessages'];
+    public decryptMessage: MetaGroupWrapper['decryptMessage'];
+    public makeSwarmSubAccount: MetaGroupWrapper['makeSwarmSubAccount'];
+    public swarmSubaccountSign: MetaGroupWrapper['swarmSubaccountSign'];
+  }
+
+  export type MetaGroupActionsType =
+    | ['init', GroupWrapperConstructor]
+    // shared actions
```
Comment (yougotwill) [inline types/groups/metagroup.d.ts:121]: New line?

File: src/utilities.cpp (hunk 22 above)
Comment (yougotwill) [reply src/utilities.cpp:202]: I mean obj.

File: types/groups/metagroup.d.ts (hunk 26 above)
Comment (yougotwill) [reply types/groups/metagroup.d.ts]: Thank you!

**Review by yougotwill — CHANGES_REQUESTED** [review]

File: src/utilities.cpp (hunk 22 above)
Comment (Bilb) [reply src/utilities.cpp:202]: yeah obj will always have those props, guaranteed

File: types/groups/metagroup.d.ts (hunk 31 above)
Comment (Bilb) [reply types/groups/metagroup.d.ts:121]: added

File: src/utilities.cpp (hunk 22 above)
Comment (yougotwill) [reply src/utilities.cpp:202]: Cool

**Review by yougotwill — APPROVED** [review]
LGTM 🚢

**Pull request merged**